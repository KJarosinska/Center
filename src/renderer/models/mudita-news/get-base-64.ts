import axios from "axios"

export const getBase64 = async (url?: string) => {
  try {
    return await axios
      .get(url, {
        responseType: "arraybuffer",
        proxy: {
          host: "localhost",
          port: 2003,
        },
      })
      .then(response => Buffer.from(response.data, "binary").toString("base64"))
  } catch (e) {
    console.log(e)
  }
}
