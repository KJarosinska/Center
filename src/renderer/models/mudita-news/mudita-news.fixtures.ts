export const newsResponse = {
  sys: { type: "Array" },
  total: 3,
  skip: 0,
  limit: 100,
  items: [
    {
      sys: {
        space: {
          sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
        },
        id: "23cbaCgEf6PC1llqI5XYBf",
        type: "Entry",
        createdAt: "2019-07-11T08:54:31.710Z",
        updatedAt: "2020-01-22T12:08:23.397Z",
        environment: {
          sys: { id: "master", type: "Link", linkType: "Environment" },
        },
        revision: 13,
        contentType: {
          sys: { type: "Link", linkType: "ContentType", id: "newsItem" },
        },
        locale: "en-US",
      },
      fields: {
        category: "Forum",
        title: "Hot discussion",
        content: "One feature I would love to see added",
        image: {
          sys: {
            type: "Link",
            linkType: "Asset",
            id: "6j1K3VOLf3aAdijTuObagd",
          },
        },
        communityLink:
          "https://forum.mudita.com/t/one-feature-i-would-love-to-see-added/299",
        link:
          "https://forum.mudita.com/t/one-feature-i-would-love-to-see-added/299",
        discussionId: "299",
      },
    },
    {
      sys: {
        space: {
          sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
        },
        id: "3RU0ukJVFGhzT61mXB4xn9",
        type: "Entry",
        createdAt: "2019-07-11T09:48:32.097Z",
        updatedAt: "2020-01-13T16:04:08.966Z",
        environment: {
          sys: { id: "master", type: "Link", linkType: "Environment" },
        },
        revision: 28,
        contentType: {
          sys: { type: "Link", linkType: "ContentType", id: "newsItem" },
        },
        locale: "en-US",
      },
      fields: {
        category: "Blog",
        title: "Featured Article",
        content: "Attention: The precious currency weâ€™re giving away for free.",
        image: {
          sys: {
            type: "Link",
            linkType: "Asset",
            id: "7qdkPb6OhjDR3ZRb90bcck",
          },
        },
        communityLink:
          "https://forum.mudita.com/t/attention-the-precious-currency-we-re-giving-away-for-free/1027",
        link:
          "https://mudita.com/community/blog/attention-the-precious-currency-were-giving-away-for-free/",
        discussionId: "1027",
      },
    },
    {
      sys: {
        space: {
          sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
        },
        id: "9YFH9NLyIvt8Z1dDEc1nf",
        type: "Entry",
        createdAt: "2019-07-11T08:58:38.928Z",
        updatedAt: "2019-10-25T10:40:45.068Z",
        environment: {
          sys: { id: "master", type: "Link", linkType: "Environment" },
        },
        revision: 24,
        contentType: {
          sys: { type: "Link", linkType: "ContentType", id: "newsItem" },
        },
        locale: "en-US",
      },
      fields: {
        category: "Blog",
        title: "Latest News",
        content:
          "Our campaign on Kickstarter was a success, reaching 262% of the initial goal!",
        image: {
          sys: {
            type: "Link",
            linkType: "Asset",
            id: "6mdeQ9Pm5Kgsek3x8Pt7Yh",
          },
        },
        communityLink:
          "https://forum.mudita.com/t/mudita-pure-kickstarter-campaign-is-live/703",
        link:
          "https://mudita.com/community/blog/the-kickstarter-campaign-of-mudita-pure-is-finished/",
        discussionId: "703",
      },
    },
  ],
  includes: {
    Asset: [
      {
        sys: {
          space: {
            sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
          },
          id: "6j1K3VOLf3aAdijTuObagd",
          type: "Asset",
          createdAt: "2019-09-24T14:33:22.922Z",
          updatedAt: "2019-09-24T14:33:22.922Z",
          environment: {
            sys: { id: "master", type: "Link", linkType: "Environment" },
          },
          revision: 1,
          locale: "en-US",
        },
        fields: {
          title: "MuditaOS",
          file: {
            url:
              "//images.ctfassets.net/isxmxtc67n72/6j1K3VOLf3aAdijTuObagd/5c406b7a69a6806003aff0a4b8cb5493/MuditaOS.jpg",
            details: { size: 77166, image: { width: 1352, height: 1079 } },
            fileName: "MuditaOS.jpg",
            contentType: "image/jpeg",
          },
        },
      },
      {
        sys: {
          space: {
            sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
          },
          id: "6mdeQ9Pm5Kgsek3x8Pt7Yh",
          type: "Asset",
          createdAt: "2019-10-24T15:22:08.851Z",
          updatedAt: "2019-10-24T15:23:03.740Z",
          environment: {
            sys: { id: "master", type: "Link", linkType: "Environment" },
          },
          revision: 2,
          locale: "en-US",
        },
        fields: {
          title: "The Kickstarter campaign of Mudita Pure was a success",
          description: "The Kickstarter campaign of Mudita Pure was a success",
          file: {
            url:
              "//images.ctfassets.net/isxmxtc67n72/6mdeQ9Pm5Kgsek3x8Pt7Yh/2a946227e8a874a8dd047b0497309d4c/final.jpg",
            details: { size: 286220, image: { width: 1440, height: 1080 } },
            fileName: "final.jpg",
            contentType: "image/jpeg",
          },
        },
      },
      {
        sys: {
          space: {
            sys: { type: "Link", linkType: "Space", id: "isxmxtc67n72" },
          },
          id: "7qdkPb6OhjDR3ZRb90bcck",
          type: "Asset",
          createdAt: "2020-01-13T15:23:22.497Z",
          updatedAt: "2020-01-13T15:23:22.497Z",
          environment: {
            sys: { id: "master", type: "Link", linkType: "Environment" },
          },
          revision: 1,
          locale: "en-US",
        },
        fields: {
          title: "Attention",
          file: {
            url:
              "//images.ctfassets.net/isxmxtc67n72/7qdkPb6OhjDR3ZRb90bcck/67c404e8e737a86ed743569ebb39f761/attention.jpg",
            details: { size: 1493422, image: { width: 1640, height: 1324 } },
            fileName: "attention.jpg",
            contentType: "image/jpeg",
          },
        },
      },
    ],
  },
}
