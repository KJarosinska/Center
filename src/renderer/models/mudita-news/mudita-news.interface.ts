import { ContentTypeLink } from "contentful"

export interface NewsEntry {
  category?: string
  title: string
  content: string
  updatedAt: string
  createdAt: string
  image?: {
    sys?: ContentTypeLink
  }
  communityLink: string
  link: string
  newsId: string
  discussionId: string
  imageSource?: string
  imageAlt?: string
}

export interface DownloadError {
  message?: string
  name?: string
  stack?: string
}

export interface Store {
  newsIds: string[]
  newsItems: NewsEntry[]
  commentsCount: Record<string, number>
  downloadError?: DownloadError
}
