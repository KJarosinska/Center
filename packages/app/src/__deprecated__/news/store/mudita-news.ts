/**
 * Copyright (c) Mudita sp. z o.o. All rights reserved.
 * For licensing, see https://github.com/mudita/mudita-center/blob/master/LICENSE.md
 */

import { RootState } from "App/__deprecated__/renderer/store"
import { createModel } from "@rematch/core"
import { Slicer } from "@rematch/select"
import { Store } from "App/__deprecated__/news/store/mudita-news.interface"
import { sortByCreationDateInDescendingOrder } from "App/__deprecated__/news/helpers/sort-by-creation-date-in-descending-order.helpers"
import { getNews, initNews } from "App/__deprecated__/renderer/requests/get-news.request"
import { DefaultNewsItems } from "App/__deprecated__/main/default-news-item"
import { RootModel } from "App/__deprecated__/renderer/models/models"

const initialState: Store = {
  newsItems: [],
}

const muditaNews = createModel<RootModel>({
  state: initialState,
  reducers: {
    update(state: Store, payload: Partial<Store>) {
      return { ...state, ...payload }
    },
  },
  effects: (d) => {
    const dispatch = d as unknown as RootState

    return {
      async loadData(
        _: any,
        rootState: { networkStatus: { online: boolean } }
      ) {
        if (rootState.networkStatus.online) {
          const data = await initNews()
          dispatch.muditaNews.update(data)
        } else {
          const defaultNews: DefaultNewsItems = await getNews()
          dispatch.muditaNews.update(defaultNews)
        }
      },
      async updateData(data: DefaultNewsItems | { updating: boolean }) {
        dispatch.muditaNews.update(data)
      },
    }
  },
  selectors: (slice: Slicer<typeof initialState>) => ({
    newsSortedByCreationDateInDescendingOrder() {
      return slice((state) => {
        return sortByCreationDateInDescendingOrder(state.newsItems)
      })
    },
  }),
})

export default muditaNews
