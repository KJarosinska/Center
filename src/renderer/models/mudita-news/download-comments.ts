import axios from "axios"
import { NewsEntry } from "Renderer/models/mudita-news/mudita-news.interface"

export const downloadComments = async (news: any) => {
  const getCommentsCountByDiscussionId = async (
    discussionId?: string,
    newsId?: string
  ): Promise<{ newsId?: string; count: number }> => {
    const {
      data: { posts_count },
    } = await axios.get(
      `${process.env.MUDITA_COMMUNITY_URL}/t/${discussionId}.json`
    )
    return { newsId, count: posts_count }
  }

  const commentsCalls = news.map(
    ({
      discussionId,
      newsId,
    }: Partial<NewsEntry>): Promise<{
      newsId?: string
      discussionId?: string
      count: number
    }> => getCommentsCountByDiscussionId(discussionId, newsId)
  )
  const commentsCounts = await Promise.all<{
    newsId: string
    count: number
  }>(commentsCalls)

  const counts = commentsCounts.reduce((acc, { newsId, count }) => {
    acc[newsId] = count
    return acc
  }, {} as Record<string, number>)

  return {
    commentsCount: counts,
  }
}
