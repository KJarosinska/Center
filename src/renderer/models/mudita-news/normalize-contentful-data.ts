import { Asset, Entry, EntryCollection } from "contentful"
import { NewsEntry } from "Renderer/models/mudita-news/mudita-news.interface"
import { getBase64 } from "Renderer/models/mudita-news/get-base-64"
import { getCommentsCount } from "Renderer/models/mudita-news/get-comments-count"

export const normalizeContentfulData = async (
  data: EntryCollection<NewsEntry>
) => {
  const { items, includes } = data
  const news = items.map(({ fields, sys }: Entry<NewsEntry>) => {
    return {
      ...fields,
      newsId: sys.id,
      updatedAt: sys.updatedAt,
      createdAt: sys.createdAt,
      imageId: fields?.image?.sys?.id,
    }
  })
  for (const item of news) {
    const {
      fields: {
        title,
        file: { url },
      },
    } = includes.Asset.find((asset: Asset) => {
      return item?.image?.sys?.id === asset.sys.id
    })
    item.commentsCount = await getCommentsCount(item.discussionId)
    item.imageSource = await getBase64(url)
    item.imageAlt = title
  }
  return {
    newsItems: news,
  }
}
