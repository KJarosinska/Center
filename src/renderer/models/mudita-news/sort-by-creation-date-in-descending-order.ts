import { NewsEntry } from "Renderer/models/mudita-news/mudita-news.interface"

export const sortByCreationDateInDescendingOrder = (newsItems: NewsEntry[]) => {
  return newsItems.sort((firstId, secondId) => {
    return (
      Number(new Date(secondId.createdAt)) - Number(new Date(firstId.createdAt))
    )
  })
}
