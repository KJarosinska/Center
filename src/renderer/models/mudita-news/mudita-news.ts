import { Dispatch } from "Renderer/store"
import axios from "axios"
import { Entry, Asset } from "contentful"
import {
  DownloadError,
  IdItem,
  NewsEntry,
  Store,
} from "Renderer/models/mudita-news/mudita-news.interface"
import { Slicer } from "@rematch/select"
import { sortByCreationDateInDescendingOrder } from "Renderer/models/mudita-news/sort-by-creation-date-in-descending-order"

const initialState: Store = {
  newsIds: [],
  newsItems: {},
  commentsCount: {},
}

export default {
  state: initialState,
  reducers: {
    update(state: Store, payload: NewsEntry[]) {
      const newsIds = payload.map(
        (news: NewsEntry): IdItem => ({
          id: news.newsId,
          createdAt: news.createdAt,
        })
      )
      const newsItems = payload.reduce(
        (acc: Record<string, NewsEntry>, newsItem: NewsEntry) => {
          acc[newsItem.newsId] = newsItem
          return acc
        },
        {} as Record<string, NewsEntry>
      )
      return { ...state, newsIds, newsItems }
    },
    updateComments(state: Store, payload: { newsId: string; count: number }[]) {
      const counts = payload.reduce((acc, { newsId, count }) => {
        acc[newsId] = count
        return acc
      }, {} as Record<string, number>)
      return {
        ...state,
        commentsCount: {
          ...state.commentsCount,
          ...counts,
        },
      }
    },
    updateError(state: Store, payload: DownloadError) {
      return {
        ...state,
        downloadError: payload,
      }
    },
  },
  effects: (dispatch: Dispatch) => ({
    async loadData() {
      const getCommentsCountByDiscussionId = async (
        discussionId?: string,
        newsId?: string
      ): Promise<{ newsId?: string; count: number }> => {
        const {
          data: { posts_count },
        } = await axios.get(
          `${process.env.GATSBY_COMMUNITY_URL}/t/${discussionId}.json`
        )
        return { newsId, count: posts_count }
      }
      try {
        const {
          data: { items, includes },
        } = await axios.get(
          `https://cdn.contentful.com/spaces/${process.env.CONTENTFUL_SPACE_ID}/environments/master/entries/?access_token=${process.env.CONTENTFUL_ACCESS_TOKEN}&content_type=newsItem`
        )
        const news = items.map(({ fields, sys }: Entry<NewsEntry>) => {
          return {
            ...fields,
            newsId: sys.id,
            createdAt: sys.createdAt,
            imageId: fields?.image?.sys?.id,
          }
        })
        news.forEach((item: NewsEntry) => {
          const {
            fields: {
              title,
              file: { url },
            },
          } = includes.Asset.find((asset: Asset) => {
            return item?.image?.sys?.id === asset.sys.id
          })
          item.imageSource = url
          item.imageAlt = title
        })
        dispatch.muditaNews.update(news)
        const commentsCalls = news.map(
          ({
            discussionId,
            newsId,
          }: Partial<NewsEntry>): Promise<{
            newsId?: string
            discussionId?: string
            count: number
          }> => getCommentsCountByDiscussionId(discussionId, newsId)
        )
        const commentsCounts: {
          newsId: string
          count: number
        }[] = await Promise.all(commentsCalls)
        dispatch.muditaNews.updateComments(commentsCounts)
      } catch (error) {
        dispatch.muditaNews.updateError(error)
        console.error(error)
      }
    },
  }),
  selectors: (slice: Slicer<typeof initialState>) => ({
    sortedIds() {
      return slice(state => {
        return sortByCreationDateInDescendingOrder(state.newsIds)
      })
    },
  }),
}
