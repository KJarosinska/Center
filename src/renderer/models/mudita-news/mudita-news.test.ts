import moxios from "moxios"
import { init } from "@rematch/core"
import muditaNews from "Renderer/models/mudita-news/mudita-news"
import { newsResponse } from "Renderer/models/mudita-news/mudita-news.fixtures"

beforeEach(() => {
  moxios.install()
})

afterEach(() => {
  moxios.uninstall()
})

it("call is performed after dispatching effect and gets initial state", () => {
  const store = init({
    models: { muditaNews },
  })

  expect(store.getState()).toMatchInlineSnapshot(`
    Object {
      "muditaNews": Object {
        "commentsCount": Object {},
        "newsIds": Array [],
        "newsItems": Array [],
      },
    }
  `)
})

it("call is performed after dispatching effect with data", async () => {
  moxios.stubRequest(
    `https://cdn.contentful.com/spaces/${process.env.CONTENTFUL_SPACE_ID}/environments/master/entries/?access_token=${process.env.CONTENTFUL_ACCESS_TOKEN}&content_type=newsItem`,
    {
      status: 200,
      response: newsResponse,
    }
  )

  moxios.stubRequest(
    new RegExp(process.env.GATSBY_COMMUNITY_URL + "/t/\\d+\\.json"),
    {
      status: 200,
      response: {
        posts_count: "10",
      },
    }
  )

  const store = init({
    models: { muditaNews },
  })
  await store.dispatch.muditaNews.loadData()

  expect(store.getState()).toMatchSnapshot()
})
