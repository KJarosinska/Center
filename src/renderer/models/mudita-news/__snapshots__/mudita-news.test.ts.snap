// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call is performed after dispatching effect with data 1`] = `
Object {
  "muditaNews": Object {
    "commentsCount": Object {
      "1027": "10",
      "299": "10",
      "703": "10",
    },
    "newsIds": Array [
      "299",
      "1027",
      "703",
    ],
    "newsItems": Object {
      "1027": Object {
        "category": "Blog",
        "communityLink": "https://forum.mudita.com/t/attention-the-precious-currency-we-re-giving-away-for-free/1027",
        "content": "Attention: The precious currency weâ€™re giving away for free.",
        "discussionId": "1027",
        "image": Object {
          "sys": Object {
            "id": "7qdkPb6OhjDR3ZRb90bcck",
            "linkType": "Asset",
            "type": "Link",
          },
        },
        "imageAlt": "Attention",
        "imageId": "7qdkPb6OhjDR3ZRb90bcck",
        "imageSource": "//images.ctfassets.net/isxmxtc67n72/7qdkPb6OhjDR3ZRb90bcck/67c404e8e737a86ed743569ebb39f761/attention.jpg",
        "link": "https://mudita.com/community/blog/attention-the-precious-currency-were-giving-away-for-free/",
        "title": "Featured Article",
      },
      "299": Object {
        "category": "Forum",
        "communityLink": "https://forum.mudita.com/t/one-feature-i-would-love-to-see-added/299",
        "content": "One feature I would love to see added",
        "discussionId": "299",
        "image": Object {
          "sys": Object {
            "id": "6j1K3VOLf3aAdijTuObagd",
            "linkType": "Asset",
            "type": "Link",
          },
        },
        "imageAlt": "MuditaOS",
        "imageId": "6j1K3VOLf3aAdijTuObagd",
        "imageSource": "//images.ctfassets.net/isxmxtc67n72/6j1K3VOLf3aAdijTuObagd/5c406b7a69a6806003aff0a4b8cb5493/MuditaOS.jpg",
        "link": "https://forum.mudita.com/t/one-feature-i-would-love-to-see-added/299",
        "title": "Hot discussion",
      },
      "703": Object {
        "category": "Blog",
        "communityLink": "https://forum.mudita.com/t/mudita-pure-kickstarter-campaign-is-live/703",
        "content": "Our campaign on Kickstarter was a success, reaching 262% of the initial goal!",
        "discussionId": "703",
        "image": Object {
          "sys": Object {
            "id": "6mdeQ9Pm5Kgsek3x8Pt7Yh",
            "linkType": "Asset",
            "type": "Link",
          },
        },
        "imageAlt": "The Kickstarter campaign of Mudita Pure was a success",
        "imageId": "6mdeQ9Pm5Kgsek3x8Pt7Yh",
        "imageSource": "//images.ctfassets.net/isxmxtc67n72/6mdeQ9Pm5Kgsek3x8Pt7Yh/2a946227e8a874a8dd047b0497309d4c/final.jpg",
        "link": "https://mudita.com/community/blog/the-kickstarter-campaign-of-mudita-pure-is-finished/",
        "title": "Latest News",
      },
    },
  },
}
`;
